core/fast_string.rs

```diff
@@ -0,0 +1,243 @@
+    }
+  }
+
+  /// Create a v8 string from this [`FastString`]. If the string is static and contains only ASCII characters,
+  /// an external one-byte static is created.
+  pub fn v8<'a>(
+    &self,
+    scope: &mut v8::HandleScope<'a>,
+  ) -> v8::Local<'a, v8::String> {
+    match self.try_static_ascii() {
+      Some(s) => v8::String::new_external_onebyte_static(scope, s).unwrap(),
+      None => {
+        v8::String::new_from_utf8(scope, self.as_bytes(), NewStringType::Normal)
+          .unwrap()
+      }
+    }
+  }
+
+  /// Truncates a `ModuleCode`] value, possibly re-allocating or memcpy'ing. May be slow.
+  pub fn truncate(&mut self, index: usize) {
+    match self {
+      Self::Static(b) => *self = Self::Static(&b[..index]),
+      Self::StaticAscii(b) => *self = Self::StaticAscii(&b[..index]),
+      Self::Owned(b) => *self = Self::Owned(b[..index].to_owned().into()),
+      // We can't do much if we have an Arc<str>, so we'll just take ownership of the truncated version
+      Self::Arc(s) => *self = s[..index].to_owned().into(),
+    }
+  }
+}
+
+impl Hash for FastString {
+  fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+    self.as_str().hash(state)
+  }
+}
+
+impl AsRef<str> for FastString {
+  fn as_ref(&self) -> &str {
+    self.as_str()
+  }
+}
+
+impl Borrow<str> for FastString {
+  fn borrow(&self) -> &str {
+    self.as_str()
+  }
+}
+
+impl Debug for FastString {
+  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+    Debug::fmt(self.as_str(), f)
+  }
+}
+
+impl Default for FastString {
+  fn default() -> Self {
+    Self::StaticAscii("")
+  }
+}
+
+impl PartialEq for FastString {
+  fn eq(&self, other: &Self) -> bool {
+    self.as_bytes() == other.as_bytes()
+  }
+}
+
+impl Eq for FastString {}
+
+/// [`FastString`] can be make cheaply from [`Url`] as we know it's owned and don't need to do an
+/// ASCII check.
+impl From<Url> for FastString {
+  fn from(value: Url) -> Self {
+    let s: String = value.into();
+    s.into()
+  }
+}
+
+/// [`FastString`] can be make cheaply from [`String`] as we know it's owned and don't need to do an
+/// ASCII check.
+impl From<String> for FastString {
+  fn from(value: String) -> Self {
+    FastString::Owned(value.into_boxed_str())
+  }
+}
+
+/// [`FastString`] can be make cheaply from [`Arc<str>`] as we know it's shared and don't need to do an
+/// ASCII check.
+impl From<Arc<str>> for FastString {
+  fn from(value: Arc<str>) -> Self {
+    FastString::Arc(value)
+  }
+}
+
+/// Include a fast string in the binary. This string is asserted at compile-time to be 7-bit ASCII for optimal
+/// v8 performance.
+#[macro_export]
+macro_rules! include_fast_string {
+  ($file:literal) => {
+    $crate::FastString::ensure_static_ascii(include_str!($file))
+  };
+}
+
+/// Include a fast string in the binary from a string literal. This string is asserted at compile-time to be
+/// 7-bit ASCII for optimal v8 performance.
+#[macro_export]
+macro_rules! fast {
```

Comment:
```suggestion
macro_rules! ascii_str {
```
I had a few head-scratches in the files above before I got here to figure out what `fast!()` was supposed to mean. I suggest to make it more descriptive - I think `ascii_str!()` will better communicate the intent and boundaries here.